### PR TITLE
feat: variables

### LINKED ISSUE #13: Variables and Mutability
Variables, mutable variables, shadowing, constants.

Can (even should!) take inspiration from 
https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html

Estimated time: 1 hour

### PR DESCRIPTION
**Pull request by enitrat**
Closes #13 

2 hours as expected

### CODE REVIEW

File: src/ch02-01-variables-and-mutability.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+## Variables and Mutability
+
+Cairo uses an immutable memory model, meaning that once a memory cell is written to, it can't be overwritten but only read from. To reflect this immutable memory model, variables in Cairo are immutable by default.
+However, the language abstracts this model and gives you the option to make your variables mutable. Let’s explore how and why Cairo enforces immutability, and how you can make your variables mutable.
+
+When a variable is immutable, once a value is bound to a name, you can’t change
+that value. To illustrate this, generate a new project called _variables_ in
+your _cairo_projects_ directory by using `scarb new variables`.
```
Comment (fricoben) [inline src/ch02-01-variables-and-mutability.md:8]: Did you previously explained what is `scarb` anon ? Cause idk what it is

File: src/ch02-01-variables-and-mutability.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+first variable is _shadowed_ by the second, which means that the second
+variable is what the compiler will see when you use the name of the variable.
+In effect, the second variable overshadows the first, taking any uses of the
+variable name to itself until either it itself is shadowed or the scope ends.
+We can shadow a variable by using the same variable’s name and repeating the
+use of the `let` keyword as follows:
+
+<span class="filename">Filename: src/lib.cairo</span>
+
+```swift
```
Comment (fricoben) [inline src/ch02-01-variables-and-mutability.md:152]: swift ?

File: src/ch02-01-variables-and-mutability.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+using the `let` keyword. By using `let`, we can perform a few transformations
+on a value but have the variable be immutable after those transformations have
+been completed.
+
+Another distinction between `mut` and shadowing is that when we use the `let` keyword again,
+we are effectively creating a new variable, which allows us to change the type of the
+value while reusing the same name. As mentioned before, variable shadowing and mutable variables
+are equivalent at the lower level.
+The only difference is that by shadowing a variable, the compiler will not complaing
+if you change its type. For example, say our program performs a type conversion between the
```
Comment (fricoben) [inline src/ch02-01-variables-and-mutability.md:200]: Good note 🫡

File: src/ch02-01-variables-and-mutability.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+use traits::Into;
+fn main() {
+    let x = 2_u64;
+    x.print();
+    let x = x.into(); // converts x to a felt.
+    x.print()
+}
+```
+
+The first `x` variable has a u64 type while the second 'x' variable has a felt252 type.
```
Comment (fricoben) [inline src/ch02-01-variables-and-mutability.md:214]: the second `x` instead of the second 'x'

`felt252` instead of felt252

**Review by fricoben — CHANGES_REQUESTED** [review]
Looks good, have some questions and some small changes to make.

File: src/ch02-01-variables-and-mutability.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
+## Variables and Mutability
+
+Cairo uses an immutable memory model, meaning that once a memory cell is written to, it can't be overwritten but only read from. To reflect this immutable memory model, variables in Cairo are immutable by default.
+However, the language abstracts this model and gives you the option to make your variables mutable. Let’s explore how and why Cairo enforces immutability, and how you can make your variables mutable.
```
[the remark this answers is not in the document]
Comment (enitrat) [reply src/ch02-01-variables-and-mutability.md:4]: no the `mut` concept is new

File: src/ch02-01-variables-and-mutability.md (hunk 1 above)
Comment (enitrat) [reply src/ch02-01-variables-and-mutability.md:8]: yes, in the chapter 1.2

File: src/ch02-01-variables-and-mutability.md (hunk 2 above)
Comment (enitrat) [reply src/ch02-01-variables-and-mutability.md:152]: rust has weird syntax highlighting when used with single quotes

**Comment by hurrikaanig** [comment]
Looks good to me !

**Review by hurrikaanig — APPROVED** [review]

**Comment by enitrat** [comment]
@fricoben I think i'm going to keep the part about immutable variables like it is currently. I think a lengthy explanation is required for readers not familiar to Rust

**Pull request merged**